File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(timer){
  if (timer==undefined)
    return "You forgot to set the timer."
  else if (timer==0)
    return "Lasagna is done."
  else
    return "Not done, please wait."
}

export function preparationTime(layers,timePrep=2){
  return layers.length*timePrep
}

export function quantities(layer){
  const qty={
    noodles:0,
    sauce:0
  }
  for(let i=0;i<layer.length;i++){
    if (layer[i]=="sauce")
      qty.sauce+=0.2
    else if (layer[i]=="noodles")
      qty.noodles+=50
  }
  return qty
}

export function addSecretIngredient(friendsList,myList){
  myList.push(friendsList[friendsList.length-1])
}

export function scaleRecipe(recipe,value){
  let copyRecipe = {...recipe}
  for (let key in copyRecipe)
    copyRecipe[key]=copyRecipe[key]*value/2
  return copyRecipe
}

File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"

namespace lasagna_master {

}  

int preparationTime(const std::vector<std::string>& layers, const int& prep_time) {
    return prep_time*layers.size();
}
int preparationTime(const std::vector<std::string>& layers) {
    return 2*layers.size();
}
lasagna_master::amount quantities(const std::vector<std::string>& layers) {
    lasagna_master::amount answer;
    answer.noodles=0;
    answer.sauce=0.0;

    for( std::string layer : layers ) {
        if (layer == "noodles")
            answer.noodles+=50;
        if (layer == "sauce")
            answer.sauce+=0.2;
    }
    
    return answer;
}

void addSecretIngredient(std::vector<std::string>& myList, std::vector<std::string> friendsList){
    myList.back()=friendsList.back();
}

void addSecretIngredient(std::vector<std::string>& myList, std::string secretIngredient) {
    myList.back()=secretIngredient;
}

std::vector<double> scaleRecipe(const std::vector<double> quantities, int portions) {
    std::vector<double> new_quantities;
    for (long unsigned int i=0; i<quantities.size(); i++) {
        new_quantities.push_back(quantities.at(i)*portions/2.0);
    }
    return new_quantities;
}